teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,167 @@
+---
+status: implementable
+title: Unified Retries for TaskRuns and CustomRuns
+creation-date: '2022-09-08'
+last-updated: '2022-09-21'
+authors:
+- '@XinruZhang'
+---
+
+# TEP-0121: Unified Retries for TaskRuns and CustomRuns
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Alternatives](#alternatives)
+- [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Retries is designed inconsistently between TaskRun and CustomRun, see the [Motivation](#motivation) section for details. In this TEP, we propose to align the inconsistency by
+1. solidifying `TimeOut` is for each retry attempt, updating [TEP-069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) and [docs/runs.md](https://github.com/tektoncd/pipeline/blob/0dd9f9719177054bdf60928ce9b4aab4b85af5bd/docs/runs.md) to reflect the alignment.
+2. migrating `Retries` implementation to TaskRun reconciler.
+
+## Motivation
+Two distinct issues motivate this design:
+- [TaskRun retries: Improve separation of concerns between PipelineRun and TaskRun reconciler](https://github.com/tektoncd/pipeline/issues/5248)
+- [TEP-0114: Clarify the Behavior of Custom Task Retries](https://github.com/tektoncd/pipeline/pull/5393)
+
+Some important details of the current implementation:
+
+| Runtime Object | Retriable Independently? | Retriable in Pipeline? | Timeout for Per-retry? | Note |
+|:---|:---:|:---:|:---:|:---:|
+|**TaskRun**| No | Yes | Yes | Implemented in PipelineRun reconciler|
+|**CustomRun**| Yes| Yes | Yes and No | Implemented in CustomRun reconciler|
+
+`TimeOut` in Custom Task Run is designed (see [TEP-0069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md)) for constraining the running time of all retry attempts, while being implemented (see the [implementation](https://github.com/tektoncd/pipeline/blob/e1fd72720652fa5a1b7f811a6ce1d6c943be11a6/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L198)) for each retry attempt.
+
+And CustomRun is retriable on its own, while TaskRun can only be retried within a PipelineRun.
+
+### Goals
+1. Align the behavior of `Retries` between TaskRun and CustomRun.
+2. Retry TaskRun independently.
```

Comment:
What is the use case of being able to retry `taskRun` independently? What is the driving factor for creating just a `taskRun` outside of `pipeline`? I am trying to recollect and gather any such use case. I have hardly seen a requirement of creating just a `taskRun`. I can think that it could be beneficial for testing purpose, but if I am eventually making that task part of a pipeline, its easier to create a pipeline to begin with. Do we know if our catalog is using `taskRun` for their testing purpose?

Reply:
Thanks for the comment Priti!

We do have standalone TaskRuns use cases:
- https://github.com/tektoncd/catalog/tree/main/task/send-to-webhook-slack/0.1
- https://github.com/tektoncd/catalog/tree/main/task/sendmail/0.1

It sounds weird to me If we believe all custom runs need to support the `retries` field while we don't necessarily need to support retries for all Task Runs.

IMHO the reason why we didn't see a lot of use cases (i.e. feature request iessues) is creating feature requests is not an effortless action. People may not bother to create feature requests for stand-alone TaskRuns potentially because they can actually write scripts to retry them. By supporting retries natively, we can group their retry attempts all together, saving them times on writing scripts etc.

Transient errors are everywhere especially in the Cloud Environment, services can be down for a short period of time making the entire TaskRun fails. I truly believe retry is something necessary for TaskRuns to avoid flaky errors.

https://learn.microsoft.com/en-us/azure/architecture/best-practices/transient-faults#why-do-transient-faults-occur-in-the-cloud explained how common the transient errors are in the Cloud env. It deepened my understanding about the nature of Cloud Services and how to make things more stable. Hope it to be helpful.

Reply:
Thanks Andrea for the use cases. That's true, users can always write a sinle-task pipeline in order to retry a Task. Therefore making TaskRun retriable on its own should not be the reason to introduce the new `retries` field for TaskRun.